Replacing PyNE with radioactivedecay for radionuclide decay calculations

lemointm opened the pull request:
<!--- Provide a general summary of your changes in the title above -->
This PR is designed to replace the functionality of PyNE in TARDIS with the python package `radioactivedecay`. 

**Description**
<!--- Describe your changes in detail -->
TARDIS uses PyNE to perform decay caluclations, and to handle nuclide name parsing. The main change will be with handling nuclide decays in `tardis/io/decay.py` with `radioactivedecay`, and to replace the usage of PyNE's `nucname` function with the `radioactivedecay.Nuclide` class' name parsing functionality throughout TARDIS. Nuclide decays will now use the `radioactivedecay.Inventory` objects and the `decay()` method.

**Motivation and context**
<!--- Why is this change required? What problem does it solve? Link issues here -->
This PR is meant to solve problems outlined in issue [#1654](https://github.com/tardis-sn/tardis/issues/1654).

**How has this been tested?**
- [ ] Testing pipeline.
- [ ] Other. <!--- please describe how you tested your changes, `pytest` flags used, etc. -->

**Examples**
<!-- If appropriate, link notebooks, screenshots and other demo stuff -->

**Type of change**
<!--- Put an `x` in all the boxes that apply -->
- [x] Bug fix. <!-- non-breaking change which fixes an issue -->
- [ ] New feature. <!-- non-breaking change which adds functionality -->
- [ ] Breaking change. <!-- fix or feature that would cause existing functionality to not work as expected -->
- [ ] None of the above. <!-- please describe -->

**Checklist**
<!--- Put an `x` in all the boxes that apply -->
- [ ] My change requires a change to the documentation.
    - [ ] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [ ] I have assigned and requested two reviewers for this pull request.

andrewfullard commented on tardis/io/decay.py:
```diff
@@ -66,10 +91,28 @@ def to_materials(self):
+            nuclear_symbol = f"{Z_to_elem(atomic_number)}{mass_number}"
+            for i in range(len(self.columns)):
+                comp_dicts[i][nuclear_symbol] = abundances[i]
+        return [Inventory(comp_dict, "Kg") for comp_dict in comp_dicts]
```
`Kg` should be `kg`

andrewfullard commented:
Need to format to `black` standards

wkerzendorf commented:
@lemointm the tests are failing - can you make sure it reruns.

lemointm commented:
Just pushed a commit for formatting, I will check the other tests next

afloers commented:
pyne also needs to be replaced in the tests

ImportError while loading conftest '/home/vsts/work/1/s/tardis/tardis/conftest.py'.
tardis/__init__.py:15: in <module>
    import pyne.data
E   ModuleNotFoundError: No module named 'pyne'

epassaro commented:
There are several references to `pyne` in the `tardis` package, use Ctrl+Shift+F on VSCode to find them.

andrewfullard commented on tardis/util/base.py:
```diff
@@ -600,6 +600,8 @@ def is_notebook():
 
 
 if is_notebook():
+    import tqdm.notebook
```
Why is this change in this PR?

lemointm replied:
That was a mistake, just a temp fix to fix a problem on my machine. I'll revert the file in a new commit.

lemointm commented:
There's an ongoing test failure where `radioactivedecay` says Ni-58 doesn't exist, this is a small issue with the package due to Ni-58 not being a progeny of any of the nuclides in the dataset (since it's produced by spontaneous fission only). I will try to get this fixed in `radioactivedecay` asap.

andrewfullard commented on tardis/io/decay.py:
```diff
@@ -20,56 +21,57 @@ def __init__(self, *args, **kwargs):
             for i in range(len(self.columns)):
                 comp_dicts[i][nuclear_symbol] = abundances[i]
-        return [material.Material(comp_dict) for comp_dict in comp_dicts]
+        return [Inventory(comp_dict, "kg") for comp_dict in comp_dicts]
```
We use cgs units in tardis. Can you make this grams?

lemointm replied:
Since we use mass fractions in TARDIS, converting from grams to kgs should be unnecessary

lemointm replied:
Turns out `radioactivedecay` can handle grams - the newest commit has these changes

andrewfullard commented on tardis/model/base.py:
```diff
@@ -581,12 +582,18 @@ def from_csvy(cls, config):
             validate_dict(csvy_model_config, schemapath=csvy_schema_file)
         )
 
+        nuclides_hyphenless = np.char.replace(DEFAULTDATA.nuclides, "-", "")
```
You created this variable in another place too. It would be better defined in a utility location for access in multiple locations.

andrewfullard commented on tardis/visualization/widgets/custom_abundance.py:
```diff
@@ -951,8 +952,11 @@ def input_symb_eventhandler(self, obj):
         try:
-            if nucname.iselement(element_symbol_string) or nucname.isnuclide(
-                element_symbol_string
+            nuclides_hyphenless = np.char.replace(DEFAULTDATA.nuclides, "-", "")
```
See here as well

andrewfullard commented on tardis/visualization/widgets/custom_abundance.py:
```diff
@@ -951,8 +952,11 @@ def input_symb_eventhandler(self, obj):
+            if (
+                element_symbol_string in Z_DICT.values()
+                or element_symbol_string in DEFAULTDATA.nuclides
+                or element_symbol_string in nuclides_hyphenless
```
This also seems to be repeated code that showed up earlier.

lemointm commented:
@andrewfullard I added a util to tardis to perform those checks that I had rewritten in a few different places. The code should be a little cleaner now.

lemointm commented:
> @andrewfullard I added a util to tardis to perform those checks that I had rewritten in a few different places. The code should be a little cleaner now.

I spoke too soon, I just discovered that `radioactivedecay` can input grams, commit coming shortly!

lemointm commented:
There also was an error where `Inventory.mass_fractions()` wasn't working properly since each nuclide is stored in a different Inventory. This was fixed by switching to `Inventory.masses('g')` instead, with an update to having inventories stored in grams.

lemointm commented:
Now there should be only the Ni-58 test error (for now). I will give an update once the fixes are made to `radioactivedecay`.

lemointm commented:
After the newest release (0.4.12) to `radioactivedecay`, all tests pass. I marked the code for review, please let me know if there are any other changes anyone would like to see.

Rodot- commented:
Can you update the astropy version to >=4.1 to remove the distutils warnings?

lemointm commented:
> Can you update the astropy version to >=4.1 to remove the distutils warnings?

I also added radioactivedecay>=0.4.12 to ensure that the latest version with Ni-58 is installed.

lemointm commented:
There still seems to be an issue with astroPy, now with importing `astropy.modeling.blackbody`. The documentation mentions that the module are deprecated as of version 4.0.

andrewfullard commented:
Reminder to revert the astropy version change in this PR so we can merge it

epassaro commented:
Also, please try to build the documentation before merging!

epassaro commented:
Black check is failing.

andrewfullard commented:
> Black check is failing.

#1940 is the reason

epassaro commented:
> > Black check is failing.
> 
> #1940 is the reason

Thanks, I think we should update `black` in https://github.com/tardis-sn/tardis/blob/master/.github/workflows/black-check.yml instead of doing in the env file.

lemointm commented:
I just reverted the astropy version, and ran the docs on my machine - everything looks good!

epassaro commented on tardis_env3.yml:
```diff
@@ -14,7 +14,7 @@ dependencies:
   - numba=0.53
   - numexpr
-  - pyne=0.7=nomoab_openmc*  # Issue #1537
+  - radioactivedecay>=0.4.12
```
Is a good idea to pin the bugfix? Maybe just minor version it's ok, what do you think @lemointm ?

lemointm replied:
I talked to Alex about this, we think that in the future the functionality TARDIS needs won't get deprecated

epassaro replied:
but that shouldn't affect the bugfix number, right?

lemointm replied:
I see what you're saying, sorry I misunderstood. The "bugfix" added some functionality, so it needs to be pinned... which is not the standard but that's how it ended up I suppose
